### ISSUE TITLE
Windows stdio handles can be null

### ISSUE THREAD

**Issue by ChrisDenton** [BurntSushi/winapi-util#11]
See also: https://github.com/rust-lang/rust/issues/88576

In short, Windows stdio functions can return null to mean that there are no handles. The [code here](https://github.com/BurntSushi/winapi-util/blob/8f291c9da272b24699f31e22c521404255462642/src/win.rs#L132) assumes stdio handles are always valid `File` handles. While this is technically an incorrect assumption, it worked out ok until Rust's new "Safe I/O" code was implemented.

A workaround will be implemented upstream but I thought there should be an issue here too, since unfortunately it reached stable, causing `env_logger` to break.

**Comment by BurntSushi** [comment]
Yeah, this impacts termcolor too: https://github.com/BurntSushi/termcolor/issues/53

When stdio handles are null on Windows, what does `println!` do?

**Comment by ChrisDenton** [comment]
Nothing appears on screen but it's successful. And `stdout().write` reports the bytes as written. I'm actually a bit surprised by that because I would have thought it was just calling the `WriteFile` API on the handle (which should fail). I wonder what it's writing to...

**Comment by BurntSushi** [comment]
That's really weird. Also, I would have expected `println!` to panic and `writeln!` to return an error given that it has a null stdout handle? Like, how does `io::stdout()` even work given that it's null? It seems like our std APIs here are presenting a lie...

**Comment by ChrisDenton** [comment]
Yeah, it really did surprise me. I had to check I wasn't misunderstanding something. From a quick investigation it seems like stdio swallows [`ebadf` errors](https://github.com/rust-lang/rust/blob/68b554e6af18726fe6fa8de2134c59c441e0b019/library/std/src/io/stdio.rs#L189) (I'd assume "ebadf" is a nix-ism) and pretends they were successful. On Windows `is_ebadf` [returns true for `INVALID_HANDLE_VALUE`](https://github.com/rust-lang/rust/blob/e1e9319d93aea755c444c8f8ff863b0936d7a4b6/library/std/src/sys/windows/stdio.rs#L381) and elsewhere [null is mapped to `INVALID_HANDLE_VALUE`](https://github.com/rust-lang/rust/blob/e1e9319d93aea755c444c8f8ff863b0936d7a4b6/library/std/src/sys/windows/stdio.rs#L50).

So, yeah, the std APIs lie.

**Comment by ChrisDenton** [comment]
Ok, I'm told this is very much intentional because otherwise, for example, `println!` would panic when running a program that's detached from consoles (e.g. a Windows GUI program). So the stdio APIs are meant to silently fail on bad values whereas `File` is free to assumes all handles are valid (because that's the promise its API makes and is meant to be upheld).

**Comment by BurntSushi** [comment]
Silent failures don't seem great. :-/ `println!` already panics when stdout gets closed anyway.

But yeah, not quite sure what to do here.

**Comment by ChrisDenton** [comment]
Hm, I think this needs some serious design work. Obviously changing the standard library behaviour breaks programs so some new APIs might be required. I think I'd need to understand more about how this is being used in practice.

One simple thing that could be done is for `Stdout`, etc to have a `try_as_handle`. That doesn't solve the problem but at least the API indicates that there is an issue to be aware of. Maybe a higher level API would be `try_as_file_ref`. I'm not sure. And I probably shouldn't be throwing out random ideas before I've had time to consider them more. 🤷
